visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts

```diff
@@ -76,47 +73,44 @@ export class ThreeSceneService implements CodeMapPreRenderServiceSubscriber {
 		this.getMapMesh().highlightBuilding(this.highlighted, this.selected, state, this.constantHighlight)
 	}
 
-	private updateCorrectMaterial(materials: Material[], nodes: CodeMapBuilding[], select: boolean) {
-		// TODO This loops introduce performance issues
-		//  refactor it please
-		return
-		const allNodes = this.mapMesh.getNodes()
-
-		for (const node of allNodes) {
-			for (const material of materials) {
-				for (const hoveredNode of nodes) {
-					if (hoveredNode !== null) {
-						if (select) {
-							if (material.userData.id === hoveredNode.node.id) {
-								material["color"].setHex(ColorConverter.convertHexToNumber(hoveredNode.color))
-							}
-						} else if (material.userData.id === hoveredNode.node.id) {
-							material["color"].setHex(ColorConverter.convertHexToNumber("#FFFFFF"))
-						} else if (material.userData.id === node.id) {
-							material["color"].setHex(ColorConverter.convertHexToNumber("#7A7777"))
-						}
-					}
-				}
+	private selectMaterial(materials: Material[]) {
+		for (const material of materials) {
+			if (material.userData.id === this.selected.node.id) {
+				material["color"].setHex(ColorConverter.convertHexToNumber(this.selected.color))
 			}
 		}
 	}
 
 	private resetMaterial(materials: Material[]) {
-		// TODO activate the function when the performance issue in updateCorrectMaterial() has been fixed.
-		return
-		const allNodes = this.mapMesh.getNodes()
-		for (const node of allNodes) {
-			for (const material of materials) {
-				if (material.userData.id === node.id) {
-					material["color"].setHex(ColorConverter.convertHexToNumber("#FFFFFF"))
-				}
+		const selectedID = this.selected ? this.selected.node.id : -1
+		const allNodesIds = new Set(this.mapMesh.getNodes().map(({ id }) => id))
+		for (const material of materials) {
+			if (allNodesIds.has(material.userData.id) && material.userData.id !== selectedID) {
+				material["color"].setHex(ColorConverter.convertHexToNumber("#FFFFFF"))
```

Comment:
Its white? The standard hash value, but sure I can extract it into a constant

Reply:
NIT: Yes, at the moment it is white but it can change. :-)
I have to understand the full if-condition to know, what kind of color this is. Some kind of "DEFAULT_MATERIAL_COLOR" would be really nice.